### DISCUSSION TITLE
Supporting Enum types

### DISCUSSION THREAD

**Discussion by soumyamahunt** [SwiftyLab/MetaCodable#31] [Ideas]
As library is getting more adaption, it is essential to provide support for enum types as well. Currently Swift compiler's way of generating implementation:

```yaml
{
   "%caseName%": {
    "%%eachArgumentName%%": "%%eachArgumentValue%%",
    "_%%eachArgumentPositionNumer%%": "%%eachArgumentValue%%" // if no argument name
  }
}
```

For example enum with each case represented in JSON with Swift compiler generated implementation:

```swift
enum SomeEnum: Codable {
    /// ```json
    /// {
    ///   "bool": {
    ///     "_0": true
    ///   }
    /// }
    /// ```
    case bool(_ variable: Bool)
    /// ```json
    /// {
    ///   "int": {
    ///     "val": 6
    ///   }
    /// }
    /// ```
    case int(val: Int)
    /// ```json
    /// {
    ///   "string": {
    ///     "_0": "text"
    ///   }
    /// }
    /// ```
    case string(String)
    /// ```json
    /// {
    ///   "multi": {
    ///     "_0":true,
    ///     "_2":"text",
    ///     "val":5
    ///   }
    /// }
    /// ```
    case multi(_ variable: Bool, val: Int, String)
}
```

I want `MetaCodable` to behave the same way when enum case argument is provided. Following are the different behavior I have in. mind:
- For cases with `_` argument label is provided, argument name will be used rather than current way of using `_%argumentPositionNumber`. i.e. For the ` SomeEnum.bool(true)` generated JSON with `MetaCodable` implementation will be:

  ```json
  {
     "bool": {
        "variable": true
     }
  }
  ```
- For cases with single argument and no argument name the value will be directly encoded/decoded rather than being nested with `_0` key. i.e. For the ` SomeEnum.string("text")` generated JSON with `MetaCodable` implementation will be:

  ```json
  {
     "string":  "text"
  }
  ```
- For cases with multiple arguments and some arguments having no name, error will be generated to provide argument name. i.e. The ` SomeEnum.multi(_ variable: Bool, val: Int, String)` will be asked to be changed to ` SomeEnum.multi(_ variable: Bool, val: Int, strVal: String)` or ` SomeEnum.multi(_ variable: Bool, val: Int, _ strVal: String)`.
- Additionally `MetaCodable` will support following dynamic format of JSON:
  ```json
  [
    {
      "type": "on",
      "onOnlyProperty": "On!"
    },
    {
      "type": "off",
      "offOnlyProperty": "Off!"
    },
  ]
  ``` 
  being represented as Enum:

  ```swift
  enum State: Codable {
    case on(On)
    case off(Off)

    struct On: Codable {
      let onOnlyProperty: String
    }

    struct Off: Codable {
      let offOnlyProperty: String
    }
  }
  ```

Currently attaching macro to case arguments is not supported, which prevents existing features provided by `CodedAt`, `CodedIn`, `Default` and `CodedBy` being used per enum case arguments. I have raised a request for support [here](https://forums.swift.org/t/attached-macro-support-for-enum-case-arguments/67952), feel free to provide your reaction and support for this.

Please suggest if you would like anything additional be supported or any changes to be made what has been mentioned here.

**Comment by Sajjon** [comment]
Nice! I think it would be better if the JSON key was `discriminator` rather than `type`, it is more common a term I think. 

But perhaps it too, could be customized ?

**Reply by soumyamahunt** [reply]
Yes the JSON key will be customizable.

**Comment by soumyamahunt** [comment]
This has been implemented in #48, the only difference being this point:
> For cases with multiple arguments and some arguments having no name, error will be generated to provide argument name. i.e. The  SomeEnum.multi(_ variable: Bool, val: Int, String) will be asked to be changed to  SomeEnum.multi(_ variable: Bool, val: Int, strVal: String) or  SomeEnum.multi(_ variable: Bool, val: Int, _ strVal: String).

For such cases, no error will be generated by `MetaCodable`, for arguments that don't have name they will be directly decoded/encoded to container instead of being keyed.
